src/main/java/com/thealgorithms/others/StringMatchFiniteAutomata.java

```diff
@@ -1,80 +1,118 @@
+import java.util.TreeSet;
 
-// An implementation of string matching using finite automata
+/**
+ * A class to perform string matching using <a href="https://en.wikipedia.org/wiki/Finite-state_machine">finite automata</a>.
+ *
+ * @author <a href="https://github.com/prateekKrOraon">Prateek Kumar Oraon</a>
+ */
 public final class StringMatchFiniteAutomata {
+
+    // Constants
+    private static final int CHARS = 256; // Total number of characters in the input alphabet
+
+    // Finite automata table
+    private static int[][] finiteAutomata;
+
+    // Private constructor to prevent instantiation
     private StringMatchFiniteAutomata() {
     }
 
-    public static final int CHARS = 256;
-    public static int[][] fa;
-    public static Scanner scanner = null;
-
     public static void main(String[] args) {
-        scanner = new Scanner(System.in);
-        System.out.println("Enter String");
-        String text = scanner.nextLine();
-        System.out.println("Enter pattern");
-        String pat = scanner.nextLine();
+        // Scanner instance for user input
+        try (Scanner scanner = new Scanner(System.in);) {
+
+            System.out.println("Enter text:");
+            String text = scanner.nextLine();
 
-        searchPat(text, pat);
+            System.out.println("Enter pattern:");
+            String pattern = scanner.nextLine();
 
-        scanner.close();
+            Set<Integer> indexFound = searchPattern(text, pattern);
+            indexFound.forEach(System.out::println);
+        }
     }
 
-    public static void searchPat(String text, String pat) {
-        int m = pat.length();
-        int n = text.length();
+    /**
+     * Searches for the pattern in the given text using finite automata.
+     *
+     * @param text    The text to search within.
+     * @param pattern The pattern to search for.
+     */
+    public static Set<Integer> searchPattern(String text, String pattern) {
+        Set<Integer> indexFound = new TreeSet<>();
+        int patternLength = pattern.length();
+        int textLength = text.length();
+
+        // Initialize finite automata table
+        finiteAutomata = new int[patternLength + 1][CHARS];
 
-        fa = new int[m + 1][CHARS];
+        // Preprocess the pattern to create the finite automata table
+        computeFiniteAutomata(pattern, patternLength);
 
-        computeFA(pat, m, fa);
+        int state = 0; // Initial state
 
-        int state = 0;
-        for (int i = 0; i < n; i++) {
-            state = fa[state][text.charAt(i)];
+        // Process the text over the finite automata
+        for (int i = 0; i < textLength; i++) {
+            state = finiteAutomata[state][text.charAt(i)];
 
-            if (state == m) {
-                System.out.println("Pattern found at index " + (i - m + 1));
+            if (state == patternLength) {
+                indexFound.add(i - patternLength + 1);
             }
         }
+        return indexFound;
     }
```

Comment:
I think that this method should look somehow like that:
```java
public static Set<Integer> searchPattern(final String text, final String pattern) {
    final var stateTransitionTable = computeStateTransitionTable(pattern);
    FiniteAutomata finiteAutomata = new FiniteAutomata(stateTransitionTable);

    Set<Integer> indexFound = new TreeSet<>();
    for (int i = 0; i < text.length(); i++) {
        finiteAutomata.consume(text.charAt(i));

        if (finiteAutomata.getState() == pattern.length()) {
            indexFound.add(i - pattern.length() + 1);
        }
    }
    return indexFound;
}
```
The `FiniteAutomata` is a class with two `private` fields: `state` and `stateTransitionTable`. The method
`public consume(final char input)` updates the `state` according to `stateTransitionTable`.